pkg/model/components/addonmanifests/dnscontroller/remap.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dnscontroller
+
+import (
+	"fmt"
+
+	"github.com/blang/semver/v4"
+	corev1 "k8s.io/api/core/v1"
+	addonsapi "k8s.io/kops/channels/pkg/api"
+	"k8s.io/kops/pkg/featureflag"
+	"k8s.io/kops/pkg/kubemanifest"
+	"k8s.io/kops/pkg/model"
+	"k8s.io/kops/pkg/model/iam"
+)
+
+// Remap remaps the dns-controller addon
+func Remap(context *model.KopsModelContext, addon *addonsapi.AddonSpec, objects []*kubemanifest.Object) error {
```

Comment:
The challenge is that it's going to be different on each cloud, and we're going to want to do this for several pods, so this is only going to get more verbose.  I also see this as an implementation detail - we could have an Route53 proxy, for example, so even on one cloud we might end up with multiple overlays.

Reply:
I wonder if we could have a template function wrapped around the PodSpec or PodTemplateSpec which injects the things. What I'm most concerned about is having to have a per-addon block of code that knows details of the manifest such as it's a Deployment, which Deployment it is, etc. It's also invisible when looking at the manifest.

As an aside, this is unfortunately not something that Helm can do, which is likely the reason people write mutating admission hooks for this sort of thing. One disadvantage I've found with mutating admission hooks, other than their complexity, is that the pre-injection template still has to pass API validation.

Reply:
My hope (over say the 2 year horizon) is that this is just an implementation detail, and that it gets easier to use kubernetes service accounts everywhere.  For example the AWS SDK could do the service account token exchange directly; or kubernetes could issue client-certificates to pods, rendering the whole per-audience service-account concept moot.

We could mark the pod template in some way, but because we end up targeting multiple places I expect it would end up as boilerplate wrapping the template that gets ignored.  I don't mind doing that - it's likely a good idea because our current logic can't cope with multiple containers - we just don't need it right now so it's difficult to have an intuition for what to do.

I agree it would be nice to be able to see the final yaml more easily.  I think it's visible in our diffing engine, but we could render the final controllers to let people see it more clearly, or something along those lines.